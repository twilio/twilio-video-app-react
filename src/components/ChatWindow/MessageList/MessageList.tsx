import React from 'react';
import { Message } from '@twilio/conversations/lib/message';
import MessageInfo from './MessageInfo/MessageInfo';
import MessageListScrollContainer from './MessageListScrollContainer/MessageListScrollContainer';
import TextMessage from './TextMessage/TextMessage';
import useVideoContext from '../../../hooks/useVideoContext/useVideoContext';
import MediaMessage from './MediaMessage/MediaMessage';

interface MessageListProps {
  messages: Message[];
}

const getFormattedTime = (message?: Message) =>
  message?.dateCreated.toLocaleTimeString('en-us', { hour: 'numeric', minute: 'numeric' }).toLowerCase();

export default function MessageList({ messages }: MessageListProps) {
  const { room } = useVideoContext();
  const localParticipant = room!.localParticipant;

  return (
    <MessageListScrollContainer messages={messages}>
      {messages.map((message, idx) => {
        const time = getFormattedTime(message)!;
        const previousTime = getFormattedTime(messages[idx - 1]);

        // Display the MessageInfo component when the author or formatted timestamp differs from the previous message
        const shouldDisplayMessageInfo = time !== previousTime || message.author !== messages[idx - 1]?.author;

        const isLocalParticipant = localParticipant.identity === message.author;

        return (
          <React.Fragment key={message.sid}>
            {shouldDisplayMessageInfo && (
              <MessageInfo author={message.author} isLocalParticipant={isLocalParticipant} dateCreated={time} />
            )}
            {message.type === 'text' && <TextMessage body={message.body} isLocalParticipant={isLocalParticipant} />}
            {message.type === 'media' && <MediaMessage media={message.media} />}
          </React.Fragment>
        );
      })}
    </MessageListScrollContainer>
  );
}
