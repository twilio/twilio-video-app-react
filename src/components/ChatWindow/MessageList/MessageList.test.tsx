import React from 'react';
import { render } from '@testing-library/react';
import MessageList from './MessageList';

jest.mock('../../../hooks/useVideoContext/useVideoContext', () => () => ({
  room: { localParticipant: { identity: 'olivia' } },
}));

const arbitraryDate = new Date(1614635301000);

const messages: any = [
  {
    author: 'olivia',
    dateCreated: arbitraryDate,
    body: 'This is a message',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
    type: 'text',
  },
  {
    author: 'tim',
    dateCreated: arbitraryDate,
    body: 'Hi Olivia!',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX2',
    type: 'text',
  },
  {
    author: 'tim',
    dateCreated: arbitraryDate,
    body: 'That is pretty rad double line message! How did you do that?',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX3',
    type: 'text',
  },
  {
    author: 'tim',
    dateCreated: arbitraryDate,
    body: 'ðŸ˜‰',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX4',
    type: 'text',
  },
  {
    author: 'olivia',
    dateCreated: new Date(1614635361000),
    body: 'Magic',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX5',
    type: 'text',
  },
  {
    author: 'olivia',
    dateCreated: new Date(1614635361000),
    body: 'lots of magic',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX6',
    type: 'text',
  },
  {
    author: 'tim',
    dateCreated: new Date(1614635361000),
    body: 'look at this app: github.com/twilio/twilio-video-app-react. Neat huh?',
    sid: 'IMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX7',
    type: 'text',
  },
];

describe('the messageList component', () => {
  it('should render correctly', () => {
    const { container } = render(<MessageList messages={messages} />);
    expect(container).toMatchSnapshot();
  });
});
