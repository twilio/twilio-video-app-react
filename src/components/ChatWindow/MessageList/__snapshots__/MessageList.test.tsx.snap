// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the messageList component should render correctly 1`] = `
<div>
  <div
    style="padding: 0px 1.2em;"
  >
    <div
      class="makeStyles-messageInfoContainer-1"
    >
      <div>
        olivia (You)
      </div>
      <div>
        9:48 pm
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2 makeStyles-isLocalParticipant-3"
      >
        <div>
          This is a message
        </div>
      </div>
    </div>
    <div
      class="makeStyles-messageInfoContainer-1"
    >
      <div>
        tim
      </div>
      <div>
        9:48 pm
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2"
      >
        <div>
          Hi Olivia!
        </div>
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2"
      >
        <div>
          That is pretty rad double line message! How did you do that?
        </div>
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2"
      >
        <div>
          ðŸ˜‰
        </div>
      </div>
    </div>
    <div
      class="makeStyles-messageInfoContainer-1"
    >
      <div>
        olivia (You)
      </div>
      <div>
        9:49 pm
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2 makeStyles-isLocalParticipant-3"
      >
        <div>
          Magic
        </div>
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2 makeStyles-isLocalParticipant-3"
      >
        <div>
          lots of magic
        </div>
      </div>
    </div>
    <div
      class="makeStyles-messageInfoContainer-1"
    >
      <div>
        tim
      </div>
      <div>
        9:49 pm
      </div>
    </div>
    <div>
      <div
        class="makeStyles-messageContainer-2"
      >
        <div>
          look at this app: 
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="http://github.com/twilio/twilio-video-app-react"
            rel="noreferrer"
            target="_blank"
          >
            github.com/twilio/twilio-video-app-react
          </a>
          . Neat huh?
        </div>
      </div>
    </div>
  </div>
</div>
`;
